File: chapter6.hs
data Shape = Circle Float | Rect Float Float | Square Float

area :: Shape -> Float
area (Circle r) = pi * r * r
area (Rect d h) = d * h
area (Square d) = d * d

circ :: Shape -> Float
circ (Circle r) = 2 * pi * r
circ (Rect d h) = 2 * pi * (d + h)
circ (Square d) = 2 * pi * (d + d)

main :: IO ()
main =
  return ()

File: chapter6.cpp
// 2-4

class Shape {
public:
    virtual float area() = 0;
    virtual float circ() = 0;
};

const float pi = 3.14;

class Rect : public Shape {
public:
    Rect();
    float area() override
    {
        return d * h;
    };
    float circ() override
    {
        return 2 * pi * (d + h);
    };
    float d;
    float h;
};

class Circle : public Shape {
public:
    Circle();
    float area() override
    {
        return pi * r * r;
    };
    float circ() override
    {
        return 2 * pi * r;
    };
    float r;
};

class Square : public Shape {
public:
    Square();
    float area() override
    {
        return d * d;
    };
    float circ() override
    {
        return 2 * pi * (d + d);
    };
    float d;
};

// main

int main()
{
    return 0;
}
